### PR TITLE
[Ray] ray oscar deploy

### PR DESCRIPTION
**Pull request by chaokunyang**
## What do these changes do?
Support deploy oscar cluster on ray. The main changes include:
* Ray cluster setup
* Integrate ray serialization with mars
  * Now mars on ray will use ray serialization for any pickleable objects except base types and collections types.
  * Mars on ray will use ray pickle5 support for zero copy.
* Ray communication bug fix

### CODE REVIEW

File: mars/oscar/backends/ray/driver.py [hunk 1]
```diff
@@ -50,7 +50,9 @@ def setup_cluster(cls, address_to_resources: Dict[str, Dict[str, Number]]):
                 num_cpus=0,  # main pool doesn't do horse work, mark it use no cpu.
                 name=address, placement_group=pg, placement_group_bundle_index=index).remote()
             n_process = bundle_spec["CPU"]
-            ray.get(actor_handle.start.remote(address, n_process))
+            # TODO move this to a function
```
Comment (fyrestone) [inline mars/oscar/backends/ray/driver.py:53]: Maybe we can move the start actor pool logic to `RayActorBackend.create_actor_pool`? Currently, the `RayActorBackend.create_actor_pool` is empty.

File: mars/oscar/backends/ray/communication.py [hunk 2]
```diff
@@ -158,7 +160,7 @@ async def recv(self):
     async def __on_ray_recv__(self, message):
         """This method will be invoked when current process is a ray actor rather than a ray driver"""
         self._msg_recv_counter += 1
-        await self._in_queue.put(message)
+        await self._in_queue.put(deserialize(*message))
```
[the remark this answers is not in the document]
Comment (chaokunyang) [reply mars/oscar/backends/ray/communication.py:163]: There are too much mars object that can't be serialzied by ray. Those objects need to use mars serialziers, enum these types one by one and register serialziers with ray seems too much cumbersome

File: mars/oscar/backends/ray/communication.py [hunk 3]
```diff
@@ -284,3 +286,10 @@ async def connect(dest_address: str,
     @implements(Client.close)
     async def close(self):
         await super().close()
+
+
+def _gen_client_id():
+    import socket
+    import os
+    import time
+    return f"{socket.gethostname()}:{os.getpid()}:{time.time_ns()}"
```
Comment (fyrestone) [inline mars/oscar/backends/ray/communication.py:295]: Hope the `time_ns()` provides enough precision. I prefer the uuid.

File: mars/oscar/api.py [hunk 4]
```diff
@@ -79,7 +79,7 @@ def setup_cluster(address_to_resources: Dict[str, Dict[str, Number]]):
 class Actor(_Actor):
     @classmethod
     def default_uid(cls):
-        return cls.__name__
+        return cls.__name__.encode()
```
Comment (fyrestone) [inline mars/oscar/api.py:82]: Why encode the class name into bytes?

**Comment by fyrestone** [comment]
We should specify the storage backend to ray in `test_ray.py`.
- Write another `cluster.yml` with storage backends: [ray].
- Specify the config when `new_cluster` in `test_ray.py`.

File: mars/oscar/api.py (hunk 4 above)
Comment (chaokunyang) [reply mars/oscar/api.py:82]: When actor instance is created in an  actor pool, uid is encoded to bytes and used as the key of the actor instance. Will encoding to bytes make the code more consistent?

File: mars/oscar/backends/ray/communication.py (hunk 3 above)
Comment (chaokunyang) [reply mars/oscar/backends/ray/communication.py:295]: Actually `{socket.gethostname()}:{os.getpid()}` will make this process unique. Using uuid is better

File: mars/oscar/api.py (hunk 4 above)
Comment (chaokunyang) [reply mars/oscar/api.py:82]: reverted

File: mars/serialization/serializables/core.py [hunk 5]
```diff
@@ -66,6 +66,10 @@ def __init__(self, *args, **kwargs):
         for key, val in kwargs.items():
             object.__setattr__(self, key, val)
 
+    def __repr__(self):
+        values = ', '.join(['{}={!r}'.format(slot, getattr(self, slot)) for slot in self.__slots__])
```
[the remark this answers is not in the document]
Comment (chaokunyang) [reply mars/serialization/serializables/core.py:70]: added `None `

**Review by fyrestone — DISMISSED** [review]
LGTM

**Review by qinxuye — APPROVED** [review]
LGTM

**Pull request merged**